src/v/utils/timed_mutex.h

```diff
@@ -0,0 +1,182 @@
+ */
+
+#pragma once
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/semaphore.hh>
+
+#include <utility>
+
+/*
+ * This is a extension of mutex.h. The primary motivation for this class
+ * is to gather statistics on how ofter the mutex is taken (lock_counter)
+ * and the id of the top lock holder.
+ *
+ * Usage
+ * =====
+ *
+ *    ```
+ *    timed_mutex m;
+ *    return m.with("timed_mutex_example", [] { ... });
+ *    ```
+ * The first argument of `with` and `lock` is const char* lock_holder_id.
+ * timed_mutex doesn't make any copy of the string, just pass the pointer
+ * around and may store it indefinely long so it should have unlimited
+ * lifetime. From the caller perspective it means they can't really pass
+ * in a dynamic or stack-allocated string: it should really only be a string
+ * literal. This is important for performance to avoid allocating on every
+ * lock()
+ */
+
+class timed_mutex;
+
+class locked_token {
+    timed_mutex* _mutex;
+    ss::semaphore_units<> _unit;
+    bool _is_moved{false};
+    const char* _lock_holder_id;
+
+public:
+    locked_token(
+      timed_mutex* mutex,
+      const char* lock_holder_id,
+      ss::semaphore_units<>&& unit) noexcept
+      : _mutex(mutex)
+      , _lock_holder_id(lock_holder_id)
+      , _unit(std::move(unit)) {}
+    locked_token(locked_token&& token) noexcept
+      : _mutex(token._mutex)
+      , _lock_holder_id(token._lock_holder_id)
+      , _unit(std::move(token._unit)) {
+        token._is_moved = true;
+    }
+    locked_token(const locked_token&) = delete;
+    locked_token() = delete;
+
+    ~locked_token() noexcept;
+};
+
+struct lock_trace {
+    std::chrono::microseconds duration{0};
+    const char* lock_holder_id{nullptr};
+};
+
+class timed_mutex {
+public:
+    using underlying_t = ss::basic_semaphore<
+      ss::semaphore_default_exception_factory,
+      ss::timer<>::clock>;
+    using duration = typename underlying_t::duration;
+    using time_point = typename underlying_t::time_point;
+
+    explicit timed_mutex(bool should_tracing)
+      : _sem(1)
+      , _is_tracing(should_tracing) {}
+
+    ss::future<> start_tracing() {
+        return this->lock("timed_mutex:start_tracing")
+          .then([this](locked_token t) {
+              if (!this->_is_tracing) {
+                  this->_is_tracing = true;
+                  _lock_counter += 1;
+                  _acquired_at = std::chrono::steady_clock::now();
+              }
+          });
+    }
+
+    ss::future<> stop_tracing() {
+        return this->lock("timed_mutex:stop_tracing")
+          .then([this](locked_token t) { this->_is_tracing = false; });
+    }
+
+    template<typename Func>
+    auto
+    with(const char* lock_holder_id, duration timeout, Func&& func) noexcept {
+        return ss::with_semaphore(
+          _sem,
+          1,
+          timeout,
+          [this, lock_holder_id, func = std::forward<Func>(func)] {
```

Comment:
i think you probably need to mark this lambda mutable so that Func can be an r-value reference.